### PR TITLE
Use an analyzer to validate file headers

### LINKED ISSUE #20785: Validate file headers during the build
Currently the header used in source files is not validated as part of the build. An existing (third-party) analyzer rule provides the ability to combine this validation with a code fix that inserts/corrects the header as necessary. Adopting this analyzer would be helpful for a few reasons:

1. It helps ensure files have the expected headers
2. It helps Roslyn developers working on the IDE experience understand the desired behavior of additional analyzers and code fixes
3. It defers the work on this particular code style feature so we can better consider the design choices to make it relevant to a broad audience

### PR DESCRIPTION
**Pull request by sharwell**
* Enable StyleCop Analyzers file header validation for C# code
* Fix all existing file header discrepancies in C# files

Fixes #20785, but only for C# files.

### CODE REVIEW

**Comment by jaredpar** [comment]
StyleCop feels too heavy handed for this. Why not a simpler, single purpose analyzer? Using StyleCop for such a simple check feels like we're sneaking StyleCop into the code base.

**Comment by jaredpar** [comment]
Dont' understand why we're bringin in all of style cop for one simple rule. Semes very heavy weight.

File: build/Targets/Settings.props [hunk 1]
```diff
@@ -210,6 +210,10 @@
              Condition="'$(OS)' == 'Windows_NT'" />
 
   <Import Project="$(RoslynDiagnosticsPropsFilePath)" Condition="'$(UseRoslynAnalyzers)' == 'true' AND Exists('$(RoslynDiagnosticsPropsFilePath)')" />
+  <ItemGroup Condition="'$(UseRoslynAnalyzers)' == 'true' AND '$(ProjectLanguage)' == 'CSharp'">
+    <PackageReference Include="StyleCop.Analyzers" Version="$(StyleCopAnalyzersNugetPackageVersion)" PrivateAssets="all" />
```
Comment (jaredpar) [inline build/Targets/Settings.props]: Manipulation of ItemGroup should be in targets files, not props.

File: build/Targets/stylecop.json [hunk 2]
```diff
@@ -0,0 +1,13 @@
+{
+  "$schema": "https://raw.githubusercontent.com/DotNetAnalyzers/StyleCopAnalyzers/master/StyleCop.Analyzers/StyleCop.Analyzers/Settings/stylecop.schema.json",
+  "settings": {
+    "documentationRules": {
+      "xmlHeader": false,
+      "companyName": "Microsoft",
+      "copyrightText": "Copyright (c) {companyName}.  All Rights Reserved.  Licensed under the {licenseName}.  See License.txt in the project root for license information.",
```
Comment (jaredpar) [inline build/Targets/stylecop.json]: Given the current setup when I hit F5 will the debug fail to launch if the copyright is wrong? That was the #1 complaint when it came to StyleCop.

Comment (sharwell) [reply build/Targets/stylecop.json]: > Given the current setup when I hit <kbd>F5</kbd> will the debug fail to launch if the copyright is wrong? 

No. The diagnostic severity is Warning, so it'll work quietly for everything except command line and CI builds where warnings are treated as errors.

**Comment by sharwell** [comment]
> Dont' understand why we're bringin in all of style cop for one simple rule. Semes very heavy weight.

These analyzers are relatively lightweight, with performance considerations taken to the extreme. Even in cases where all rules are enabled the compilation overhead is typically very low (lower in aggregate than many other single-rule analyzers).

Also, I am only interested in discussing the merits of one rule at a time. Other rules supported by this package may be relevant to us in the future, but I prefer to discuss each individually. For now, this is one rule that I believe will make things a tiny bit easier for us and prove to not have a downside.

**Comment by kuhlenh** [comment]
@sharwell -- why not write an .editorconfig rule for this? This is actually a top style ask from our customers (an analyzer to add file header and warn if it's not there).

**Pull request closed without merging**